crypto/async/arch/async_posix.c

```diff
@@ -34,25 +39,98 @@ void async_local_cleanup(void)
 
 int async_fibre_makecontext(async_fibre *fibre)
 {
+    size_t pagesize;
+    size_t stackallocsize;
+    size_t stacksize = STACKSIZE;
+    char *stackallocaddr = NULL;
+
+# if defined(_SC_PAGE_SIZE) || defined (_SC_PAGESIZE)
+    {
+#  if defined(_SC_PAGE_SIZE)
+        long tmppgsize = sysconf(_SC_PAGE_SIZE);
+#  else
+        long tmppgsize = sysconf(_SC_PAGESIZE);
+#  endif
+        if (tmppgsize < 1)
+            pagesize = PAGE_SIZE;
+        else
+            pagesize = (size_t)tmppgsize;
+    }
+# else
+    pagesize = PAGE_SIZE;
+# endif
+    if (pagesize > stacksize)
+        stacksize = pagesize;
+
+    stackallocsize = stacksize + 2 * pagesize;
+
     fibre->env_init = 0;
     if (getcontext(&fibre->fibre) == 0) {
-        fibre->fibre.uc_stack.ss_sp = OPENSSL_malloc(STACKSIZE);
-        if (fibre->fibre.uc_stack.ss_sp != NULL) {
-            fibre->fibre.uc_stack.ss_size = STACKSIZE;
+        if (posix_memalign((void **)&stackallocaddr,
```

Comment:
not clear if posix_memalign will be available everywhere.

Reply:
>not clear if posix_memalign will be available everywhere.

Understood I had a worry over that too, but this is a posix specific implementation file, and the platforms that support posix async are already quite restricted:

```
#if defined(OPENSSL_SYS_UNIX) \
    && defined(OPENSSL_THREADS) && !defined(OPENSSL_NO_ASYNC) \
    && !defined(__ANDROID__) && !defined(__OpenBSD__)
<...>
# if _POSIX_VERSION >= 200112L
```

I picked posix_memalign() over aligned_alloc() due to it being more compatible with older systems, it should be available in posix version 200112L or later which we are already restricting on.

Reply:
Why not use mmap to get page aligned memory?

Reply:
>Why not use mmap to get page aligned memory?

Yes I did consider using mmap(), the secure heap implementation uses it. I was put off by the complications around portability of the MAP_ANON/MAP_ANONYMOUS flag. posix_memalign() being a higher level API seemed to reduce that complexity by hiding it. I can change the implementation to mmap() if you feel posix_memalign() will be a barrier to the change being accepted?